Utilities/mkpack.py

```diff
@@ -0,0 +1,307 @@
+#!/usr/bin/env python
+
+"""
+Builds a resource pack from a JSON description file.
+RebbleOS
+
+The general premise here is that there are two "and a half" output
+products from mkpack:
+
+  1) A pbpack file, which eventually goes inside of a pbz.  This gets
+     flashed in one of two ways: either through the Pebble PRF flash tools
+     (and, indirectly, from the Pebble apk or from libpebble2), or through
+     a "fake flash" tool that constructs a $PLATFORM_spi.bin for QEMU.
+
+  2) A C header file, which is included from RebbleOS.  Each resource
+     compiled by mkpack is given an enum entry, and the C header file
+     allows the resources to be references symbolically.
+
+  2.5) A dependency file, to be used by a build system.  Ideally, make
+       knows to ask mkpack to generate a dependency file; that way, a
+       pbpack is rebuilt every time either the input JSON changes, or any
+       of the resource pack inputs change.
+
+mkpack is smart enough to know to get resources in a handful of ways.  It
+can (aspirationally):
+
+  * Extract a raw resource from another resource pack (handy for
+    extracting fonts and such, before we have replacements of our own).
+  * Import a raw resource from disk (including a PNG, or a font file).
+  * Import an image from disk, converting and crushing to a palettized
+    PNG.
+  * Convert a graphic to system framebuffer format, for use as a splash
+    screen.
+"""
+
+__author__ = "Joshua Wise <joshua@joshuawise.com>"
+
+from stm32_crc import crc32
+import struct
+import json
+
+TAB_OFS = 0x0C
+RES_OFS = 0x200C
+
+def load_resource_from_pbpack(fname, resid):
+    """
+    Returns resource number |resid| from the pbpack file specified by
+    |fname|.
+    """
+    
+    with open(fname, 'r') as fd:
+        nrsrc = struct.unpack('I', fd.read(4))[0]
+        
+        for i in range(nrsrc):
+            fd.seek(TAB_OFS + i * 16)
+            (idx, ofs, sz, crc) = struct.unpack('iiiI', fd.read(16))
+            
+            if idx == resid:
+                break
+        
+        if idx != resid:
+            raise IOError("res %d not found in pbpack \"%s\"" % (resid, fname, ))
```

Comment:
I'd probably skip the trailing comma here.

Reply:
I have this weird habit of ending all of my tuples with `, `, for symmetry with the unary-product case.  I'm not sure where I learned it.  If real Python programs don't do that though, I won't either...